locked-asset/token-unstake/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+#![no_std]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+pub const DEFAULT_UNBOND_EPOCHS: u64 = 10;
+
+#[derive(ManagedVecItem, TopEncode, TopDecode, NestedEncode, NestedDecode, TypeAbi, Clone)]
+pub struct UnstakePair<M: ManagedTypeApi> {
+    pub unlock_epoch: u64,
+    pub token_payment: EsdtTokenPayment<M>,
+}
+
+#[elrond_wasm::contract]
+pub trait TokenUnstakeModule: token_send::TokenSendModule {
+    #[init]
+    fn init(&self) {
+        self.unbond_epochs().set_if_empty(DEFAULT_UNBOND_EPOCHS);
+    }
+
+    #[only_owner]
+    #[endpoint(setUnbondEpochs)]
+    fn set_unbond_epochs(&self, unbond_epochs: u64) {
+        self.unbond_epochs().set(unbond_epochs);
+    }
+
+    #[payable("*")]
+    #[endpoint(depositUserTokens)]
+    fn deposit_user_tokens(&self, caller: &ManagedAddress) {
+        let token_payment = self.call_value().single_esdt();
+        require!(
+            token_payment.token_nonce == 0,
+            "Can only unstake fungible tokens"
+        );
+        require!(
+            token_payment.amount > 0,
+            "Payment amount must be greater than 0"
+        );
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unbond_epochs = self.unbond_epochs().get();
+        let unstake_pair = UnstakePair {
+            unlock_epoch: current_epoch + unbond_epochs,
+            token_payment,
+        };
+        self.unlocked_tokens_for_user(&caller)
+            .update(|unstake_pairs| {
+                unstake_pairs.push(unstake_pair);
+            });
```

Comment:
I think we can optimize here. To store the least amount of entries, we should check if the last entry has the same `unlock_epoch`, and if yes, we can simply do `amount += ` on that entry.

Reply:
I think the check is a bit more complex as we have a managedVec of unknown size. We need to make a lot of new checks (if the vector is empty, how to get the last item and only update that one and so on), and I don't know if it is worth the extra code for that optimization.